_posts/2020-05-20-hive-acid.md

```diff
@@ -0,0 +1,317 @@
+columns.
+
+## Compactions
+The working described above with delta and delete_delta directories for each
+transaction makes the DML queries execute fast but have
+the following impact on DQL queries:
+1. Many Delta directories with small data in each directory will slow down
+execution of DQL queries. This is a known problem around
+small files where engines end up spending more time opening files than actually
+processing the data.
+2. Cross referencing all delete_delta directories to remove all deleted rows
+slows down the reads.
+
+To solve these problems, Hive compacts delta directories asynchronously at two
+levels:
+1. Minor Compaction: This compaction combines active delta directories into one
+delta directory and active delete_delta directories into one delete_delta
+directory thereby decreasing the number of small files. Limiting scope of this
+compaction to combining only delta directories keeps it fast. Minor compaction
+is automatically triggered as soon as active delta directories count reaches
+10 (configurable). This compaction creates new delta directories like
+"delta_\<start_write_id>_\<end_write_id>" where [start_write_id, end_write_id]
+gives the range of existing delta directories that we compacted. Similar naming
+convention is used for delete_delta directory.
+2. Major Compaction: Minor compaction does not work on merging base, delta and
+delete_delta directories as that requires rewriting of data with only the
+non-deleted rows, hence time consuming. This work is handled by a separate, less
+frequent and longer running, compaction called Major compaction. Major
+compaction is triggered when the total size of delta directories reaches
+10% (configurable) of the base directory size. This compaction creates a new
+Base directory.
+
+## Locks
+Hive uses shared locks to control what operations can run in parallel on
+partition/table. E.g. DML queries take a write-lock on partitions they are
+modifying while DQL queries take a read-lock on partitions it is reading.
+The read-locks taken by DQL queries prevents Hive from cleaning up the delta
+directories that have been compacted while they are being read by the query.
+
+# Changes in Presto to support Hive ACID Tables
+
+At high level, there are changes at two places in Presto to support Hive ACID
+tables: In split generation logic that runs in coordinator and in ORC reader
+that is used in workers.
+
+## Split generation
+
+1. Hive ACID State is setup in `SemiTransactionalHiveMetastore.beginQuery`,
+only for Hive ACID tables:
+    1. A new Hive transaction is opened per Query
+    2. A shared read-lock is obtained from Metastore server for the partitions
+    read in the query
+    3. A Heartbeat mechanism is set up to inform the Metastore server about
+    liveliness periodically. Frequency of heartbeats is figured out from the
+    server but can be overridden with `hive.transaction-heartbeat-interval`
```

Comment:
"server" here is a bit ambiguous. I'd say with "Metastore" or "Metastore server"